base/src/main/java/proguard/optimize/inline/IterativeInstructionVisitor.java

```diff
@@ -0,0 +1,39 @@
+package proguard.optimize.inline;
+
+import proguard.classfile.Clazz;
+import proguard.classfile.Method;
+import proguard.classfile.attribute.CodeAttribute;
+import proguard.classfile.instruction.Instruction;
+import proguard.classfile.instruction.InstructionFactory;
+import proguard.classfile.instruction.visitor.InstructionVisitor;
+
+public class IterativeInstructionVisitor {
```

Comment:
Can you describe/show a case where this class is necessary?

Reply:
The lambda usage finder looks at the program and tries to identify lambda usages, when it finds a lambda usage it will call the usage handler and it will inline the lambda. Because inlining the lambda results in a method call being replaced by a call to another method which has the lambda inlined and we could still inline lambdas into that new method because the method has multiple lambda arguments we need to stop the current visit and visit the new changed code attribute instead. This class provides a simple API to do exactly that, iterate over the instructions and restart the iteration once the code has changed.

Reply:
I can see the need for this when you have nested lambdas. But I don't understand why it's necessary when you have multiple separate lambdas in one method. Why can't you register all the changes that need to be done and apply them with the `CodeAttributeEditor` in one go?

Reply:
The lambda inliner inlines one lambda at a time so you might have a call instruction to a method, the method has for example 2 lambdas as arguments. If we inline the first lambda we create a new method that now only has one argument. If we want to inline the second lambda we have to inline that lambda into the newly created method to get a new method that now has 0 arguments. We revisit the code attribute because when finding the usages we want to find the new method so we inline into the new method. If we don't do that you will get one method that takes in lambda 1 but not lambda 2 and one method that takes in lambda 2 but not lambda 1.